From: Vivien Nicolas <vnicolas@mozilla.com>

Ensure chrome.runtime works when hosted into mozbrowser iframe containers. r=?

diff --git a/toolkit/components/extensions/MessageChannel.jsm b/toolkit/components/extensions/MessageChannel.jsm
--- a/toolkit/components/extensions/MessageChannel.jsm
+++ b/toolkit/components/extensions/MessageChannel.jsm
@@ -565,17 +565,19 @@ this.MessageChannel = {
    * @param {object} data
    * @param {nsIMessageSender|nsIMessageManagerOwner} data.target
    */
   _handleMessage(handlers, data) {
     // The target passed to `receiveMessage` is sometimes a message manager
     // owner instead of a message manager, so make sure to convert it to a
     // message manager first if necessary.
     let {target} = data;
-    if (!(target instanceof Ci.nsIMessageSender)) {
+    if (target instanceof Ci.nsIDOMHTMLIFrameElement) {
+      target = target.QueryInterface(Ci.nsIFrameLoaderOwner).frameLoader.messageManager;
+    } else if (!(target instanceof Ci.nsIMessageSender)) {
       target = target.messageManager;
     }
 
     let deferred = {
       sender: data.sender,
       messageManager: target,
     };
     deferred.promise = new Promise((resolve, reject) => {
